apps/web/src/hooks/useInlineComponent.tsx

```diff
@@ -0,0 +1,21 @@
+import { useMemo } from 'react';
+import type { ComponentType, ReactNode } from 'react';
+
+import { useDataRef } from './useDataRef';
+
+export const useInlineComponent: <T>(
+  Component: ComponentType<T>,
+  props: T & { children?: ReactNode }
+) => ComponentType<{}> = (Component, props) => {
+  const dataRef = useDataRef(props);
+
+  return useMemo(() => {
+    const componentName = Component.displayName || Component.name;
+
+    const OuterComponent = () => <Component {...dataRef.current} />;
+
+    OuterComponent.displayName = `${componentName}HOC`;
+
+    return OuterComponent;
+  }, [Component, dataRef]);
+};
```

Comment:
The new hook that creates the memoized HOC for any component. It allows to return the components from the hooks. Whenever it will be rendered the dataRef will pass the latest props. Check how it's used in the `useTemplatesStoreModal`.